### DISCUSSION TITLE
How to get verbose error messages about "circular equalities" in OMPython?

### DISCUSSION THREAD

**Discussion by Thewodoros** [OpenModelica/OpenModelica#12265] [Q&A] [answered]
Note: this is a copy of [my question](https://stackoverflow.com/questions/78341743/how-to-get-verbose-error-messages-about-circular-equalities-in-ompython) on StackOverflow. I'm not sure where I can get a reply faster. I hope it's not against the rules.

I have a modelica file (similar to the one in [this question](https://stackoverflow.com/questions/70161535/modelica-reduction-method-pantelides-error-in-electrical-due-to-missing-ground)) which fails to simulate. Here it is for reference:

`circuit.mo`
```
model circuit
	Modelica.Electrical.Analog.Basic.Resistor Resistor_2(R = 1);
	Modelica.Electrical.Analog.Sources.ConstantVoltage ConstantVoltage_3(V = 1);
	Modelica.Electrical.Analog.Basic.Resistor Resistor_4(R = 1);
	Modelica.Electrical.Analog.Basic.Inductor Inductor_5(L = 1);
	Modelica.Electrical.Analog.Sources.SineVoltage SineVoltage_6(V = 1, f = 1, phase = 1, offset = 0);
equation
	connect(Resistor_2.n, ConstantVoltage_3.p);
	connect(ConstantVoltage_3.n, Resistor_4.p);
	connect(Resistor_4.n, Inductor_5.p);
	connect(Inductor_5.n, SineVoltage_6.n);
	connect(Resistor_2.p, SineVoltage_6.p);
end circuit;
```
I try to run it from Python with code like this:

```
from OMPython import OMCSessionZMQ
omc = OMCSessionZMQ()
cmds = ['loadFile("circuit.mo")',
        'simulate(circuit)']
for cmd in cmds:
    omc.sendExpression(cmd)
```
Then I find the log file in the current directory and parse it to check for errors.

This code works OK with correct .mo files. But for this one, no log file is even created.

When I run this file with `omc -u circuit.mo`, I get a verbose error message about "circular equalities":
```
Error processing file: circuit.mo
Notification: Automatically loaded package Complex 4.0.0 due to uses annotation from Modelica.
Notification: Automatically loaded package ModelicaServices 4.0.0 due to uses annotation from Modelica.
Notification: Automatically loaded package Modelica 4.0.0 due to usage.
Error: Internal error Circular Equalities Detected for Variables:
----------------------------------
Error: Internal error IndexReduction.pantelidesIndexReduction failed! Found empty set of continuous equations. Use -d=bltdump to get more information.
Error: Internal error Transformation Module PFPlusExt index Reduction Method Pantelides failed!

# Error encountered! Exiting...
# Please check the error message and the flags.

Execution failed!
```
I would like to be able to get such messages in my Python code.

However, If I run it in OMShell-terminal, the only error message I get is `Failed to build model: circuit`.

```
>>> loadFile("circuit.mo")
true

>>> simulate(circuit)
record SimulationResult
    resultFile = "",
    messages = "Failed to build model: circuit"
end SimulationResult;

>>> translateModel(circuit)
false

>>> buildModel(circuit)
{"",""}

```
I also tried adding `getErrorString()` after each of these commands, but it always returns an empty string.

Back in the Python code, I tried saving the return values of `sendExpression()` and parsing them, but that gives me the same messages as in OMShell.

So, is it possible to get that verbose error message (as in `omc`) through `OMPython`?

**Comment by casella** [comment]
> Note: this is a copy of [my question](https://stackoverflow.com/questions/78341743/how-to-get-verbose-error-messages-about-circular-equalities-in-ompython) on StackOverflow. I'm not sure where I can get a reply faster. I hope it's not against the rules.

Of course not. StackOverflow is an independent agency 😄

**Comment by casella** [comment]
Your circuit model lacks a ground connection. Hence, the voltages are undetermined. Just add a ground component and connect it to a pin of a component.

**Reply by Thewodoros** [reply]
Sorry I didn’t make that clearer, but I know what’s wrong with the Modelica code. This code is just an example. I want to know how to get more verbose error messages in my Python code (I’m trying to write my own shell for simulating models).

**Reply by arun3688** [reply, marked as answer]
@Thewodoros can you provide your python script, are you using `ModelicaSystem` constructor in your script

**Reply by Thewodoros** [reply]
Thank you @arun3688 , looks like using `ModelicaSystem` is the right solution. For reference, the code below was run with `OMPython==3.5.1`.

My old code using `OMCSessionZMQ`:
```
from OMPython import OMCSessionZMQ
omc = OMCSessionZMQ()
cmds  = ['loadFile("circuit.mo")',
         'simulate(circuit)']
for cmd in cmds:
    ans = omc.sendExpression(cmd)
    print(ans)
```
It prints a non-descriptive error message:
```
True
{'resultFile': '', 'simulationOptions': "startTime = 0.0, stopTime = 1.0, numberOfIntervals = 500, tolerance = 1e-06, method = 'dassl', fileNamePrefix = 'circuit', options = '', outputFormat = 'mat', variableFilter = '.*', cflags = '', simflags = ''", 'messages': 'Failed to build model: circuit', 'timeFrontend': 0.0, 'timeBackend': 0.0, 'timeSimCode': 0.0, 'timeTemplates': 0.0, 'timeCompile': 0.0, 'timeSimulation': 0.0, 'timeTotal': 0.0}
```
My new code using `ModelicaSystem` (I had to use `abspath` because relative path didn't work):
```
import os
from OMPython import ModelicaSystem
mod = ModelicaSystem(os.path.abspath('circuit.mo'), 'circuit')
```
It prints the verbose message I wanted:
```
Error: Internal error Circular Equalities Detected for Variables:
----------------------------------
Error: Internal error IndexReduction.pantelidesIndexReduction failed! Found empty set of continuous equations. Use -d=bltdump to get more information.
Error: Internal error Transformation Module PFPlusExt index Reduction Method Pantelides failed!
```

**Comment by arun3688** [comment]
> from OMPython import OMCSessionZMQ
> omc = OMCSessionZMQ()
> cmds  = ['loadFile("circuit.mo")',
>          'simulate(circuit)']
> for cmd in cmds:
>     ans = omc.sendExpression(cmd)
>     print(ans)

@Thewodoros even `OMCSessionZMQ` should work, because `ModelicaSystem`  itself internally uses `sendExpression` to get the results, but it has some different approach to do simulation, what you need here in the above script is `getErrorString()` which should give you the expected error message

```
> from OMPython import OMCSessionZMQ
> omc = OMCSessionZMQ()
> cmds  = ['loadFile("circuit.mo")', 'getErrorString()',
>          'simulate(circuit)', 'getErrorString()']
> for cmd in cmds:
>     ans = omc.sendExpression(cmd)
>     print(ans)

```

**Reply by Thewodoros** [reply]
@arun3688 Thank you, that works too. However I noticed that when I use `ModelicaSystem`, it just prints the messages to stdout, and I can't find them in the `mod` object itself. I checked `vars(mod)` and the messages are not there. I also checked `dir(mod)` and I don't see any method with "error" in its name.

So I had to use `contextlib.redirect_stdout` to access them from the code.

**Reply by arun3688** [reply]
@Thewodoros Yes it just prints the messages to `stdout`
